tools/c7n_gcp/c7n_gcp/query.py

```diff
@@ -351,10 +374,88 @@ class TypeInfo(metaclass=TypeMeta):
     # cloud asset inventory type
     asset_type = None
 
+    # URN generation
+    urn_region_key = 'region'
+    # A jmespath into the resource object to find the id element of the URN.
+    # If unset, it uses the value for id.
+    urn_id_path = None
+    # It is frequent enough that the id we want for the URN is made up of one or more
+    # path segments from the id. Ids are frequently '/' delimited strings.
+    # If set, this should be an iterable of integer indices into the segments.
+    urn_id_segments = None
+    # By default the component is taken for the URN. Can be overridden by specifying
+    # a specific urn_component.
+    urn_component = None
+    # Truly global resources should override this to the empty string.
+    urn_has_project = True
+    # The location element is a zone, not a region.
+    urn_zonal = False
+
     @classmethod
     def get_metric_resource_name(cls, resource):
         return resource.get(cls.name)
 
+    @classmethod
+    def get_urns(cls, resources, project_id):
+        """Generate URNs for the resources.
+
+        A Uniform Resource Name (URN) is a URI that identifies a resource by
+        name in a particular namespace. A URN may be used to talk about a
+        resource without implying its location or how to access it.
+
+        The generated URNs can uniquely identify any given resource.
+
+        The generated URN is intended to follow a similar pattern to ARN, but be
+        specific to GCP.
+
+        gcp:<service>:<location>:<project>:<resource-type>/<resource-id>
+
+        If the region is "global" then it is omitted from the URN.
+        """
+        from pprint import pprint
+        pprint(resources)
+        return [cls._get_urn(r, project_id) for r in resources]
+
+    @classmethod
+    def _get_urn(cls, resource, project_id) -> str:
+        "Generate an URN for the resource."
+        location = cls._get_location(resource)
+        if location == "global":
+            location = ""
+        id = cls._get_id(resource)
+        if not cls.urn_has_project:
+            project_id = ""
+        # NOTE: not sure whether to use `component` or just the last part of
+        # `component` (split on '.') for the part after project
+        return f"gcp:{cls.service}:{location}:{project_id}:{cls.urn_component}/{id}"
+
+    @classmethod
+    def _get_id(cls, resource):
```

Comment:
just thinking we should prefix this one with _get_urn_id to avoid ambiguity.

Reply:
sounds good to me